hikari/impl/rest.py

```diff
@@ -1273,46 +1278,73 @@ async def _create_message(
                 embed_payload, embed_attachments = self._entity_factory.serialize_embed(embed)
                 final_attachments.extend(embed_attachments)
-                serialized_embeds.append(embed_payload)
+                serialized_embeds = [embed_payload]
+        elif embeds is not undefined.UNDEFINED:
+            update_embeds = True
+            if embeds is not None:
+                serialized_embeds = []
+                for embed in embeds:
+                    embed_payload, embed_attachments = self._entity_factory.serialize_embed(embed)
+                    final_attachments.extend(embed_attachments)
+                    serialized_embeds.append(embed_payload)
+
+        if edit:
+            # Special cases for message editing
+            if not undefined.all_undefined(mentions_everyone, mentions_reply, user_mentions, role_mentions):
+                # If specified we override the message's allowed mentions, else
+                # whatever the current value is will be used
+                body.put(
+                    "allowed_mentions",
+                    mentions.generate_allowed_mentions(mentions_everyone, mentions_reply, user_mentions, role_mentions),
+                )
+            if content is not None:
+                body.put("content", content, conversion=str)
+            else:
+                # None here means we want to remove the content from the message
+                body.put("content", None)
+            if replace_attachments:
+                # We are removing embeds and/or replacing them with new ones
+                body.put("attachments", None)
+        else:
+            body.put(
+                "allowed_mentions",
+                mentions.generate_allowed_mentions(mentions_everyone, mentions_reply, user_mentions, role_mentions),
+            )
+            body.put("content", content, conversion=str)
 
-        body.put(
-            "allowed_mentions",
-            mentions.generate_allowed_mentions(mentions_everyone, mentions_reply, user_mentions, role_mentions),
-        )
-        body.put("content", content, conversion=str)
         body.put("tts", tts)
-        body.put("embeds", serialized_embeds)
+
+        if update_embeds:
+            body.put("embeds", serialized_embeds)
 
         if final_attachments:
             form = data_binding.URLEncodedFormBuilder(executor=self._executor)
-            form.add_field("payload_json", data_binding.dump_json(body), content_type=_APPLICATION_JSON)
 
             for i, attachment in enumerate(final_attachments):
                 form.add_resource(f"file{i}", attachment)
+            return form
 
-            response = await self._request(route, form_builder=form, query=query, no_auth=no_auth)
-        else:
-            response = await self._request(route, json=body, query=query, no_auth=no_auth)
-
-        assert isinstance(response, dict)
-        return self._entity_factory.deserialize_message(response)
+        return None
```

Comment:
Instead of returning the form, could this be changed to actually do the request and then return whatever `_request` returns. The logic for deserializing the message can be moved outside. The interface will be the same as if you were doing `_request` but with it handling all the logic

Reply:
not all endpoints are putting the message body in the in the form builder under the key "payload_json" so that'd leak endpoint specific impl detail into this code which its meant to avoid

Reply:
Just what I was about to type lol 😅

Reply:
It looks like they all are tho

- https://github.com/hikari-py/hikari/pull/971/files#diff-9904e7f903a912bfe485f6f673fcd1e42da8b90786d58fb612077b2c42b68bfaR1394
- https://github.com/hikari-py/hikari/pull/971/files#diff-9904e7f903a912bfe485f6f673fcd1e42da8b90786d58fb612077b2c42b68bfaR1590
- https://github.com/hikari-py/hikari/pull/971/files#diff-9904e7f903a912bfe485f6f673fcd1e42da8b90786d58fb612077b2c42b68bfaR1921
- https://github.com/hikari-py/hikari/pull/971/files#diff-9904e7f903a912bfe485f6f673fcd1e42da8b90786d58fb612077b2c42b68bfaR1989
- https://github.com/hikari-py/hikari/pull/971/files#diff-9904e7f903a912bfe485f6f673fcd1e42da8b90786d58fb612077b2c42b68bfaR3521
- https://github.com/hikari-py/hikari/pull/971/files#diff-9904e7f903a912bfe485f6f673fcd1e42da8b90786d58fb612077b2c42b68bfaR3570

Reply:
`create_initial_response` adds message to a nested `data` attribute iirc

Reply:
I don't think its the form data that was the problem it was the actual json body